app/eth1wrap/client.go

```diff
@@ -0,0 +1,110 @@
+// Copyright © 2022-2024 Obol Labs Inc. Licensed under the terms of a Business Source License 1.1
+
+package eth1wrap
+
+import (
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+
+	"github.com/obolnetwork/charon/app/errors"
+	erc1271 "github.com/obolnetwork/charon/app/eth1wrap/generated"
+)
+
+//go:generate abigen --abi=build/IERC1271.abi --pkg=erc1271 --out=generated/erc1271.go
+
+var erc1271MagicValue = [4]byte{0x16, 0x26, 0xba, 0x7e}
+
+// NewEth1Client returns a initiliazed eth1 JSON-RPC client.
+func NewEth1Client(addr string) (*Client, error) {
+	cl := NewLazyEth1Client(addr)
+	err := cl.initializeClient()
+
+	return cl, err
+}
+
+// NewEth1Client returns an uninitialized eth1 JSON-RPC client.
+func NewLazyEth1Client(addr string) *Client {
+	return &Client{
+		eth1Cl:                 nil,
+		executionEngineAddress: addr,
+	}
+}
+
+// Client wraps a eth1 client
+type Client struct {
+	eth1ClMu               sync.RWMutex
+	eth1Cl                 *ethclient.Client
+	executionEngineAddress string
+}
+
+// isInitialized returns true if the eth1 client is initialized and false otherwise.
+func (cl *Client) isInitialized() bool {
+	cl.eth1ClMu.RLock()
+	defer cl.eth1ClMu.RUnlock()
+
+	return cl.eth1Cl != nil
+}
+
+// maybeInitializeClient initializes the eth1 client if not initialized.
+func (cl *Client) maybeInitializeClient() error {
+	if cl.executionEngineAddress == "" {
+		return errors.New("empty execution engine address")
+	}
+
+	if cl.isInitialized() {
+		return nil
+	}
+
+	cl.eth1ClMu.Lock()
+	defer cl.eth1ClMu.Unlock()
+
+	// double-check, another goroutine might have initialized in the meantime
+	if cl.eth1Cl != nil {
+		return nil
+	}
+
+	if cl.eth1Cl == nil {
+		err := cl.initializeClient()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// initializeClient initializes the eth1 client. This function is not thread safe.
+func (cl *Client) initializeClient() error {
+	eth1Cl, err := ethclient.Dial(cl.executionEngineAddress)
```

Comment:
You shall provide access to `Close()` function, otherwise the user won't be able to close it.